// @binance-chain/javascript-sdk/lib/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "BncClient", {
  enumerable: true,
  get: function get() {
    return _client.BncClient;
  }
});
Object.defineProperty(exports, "rpc", {
  enumerable: true,
  get: function get() {
    return _rpc["default"];
  }
});
Object.defineProperty(exports, "Transaction", {
  enumerable: true,
  get: function get() {
    return _tx["default"];
  }
});
exports.utils = exports.types = exports.crypto = exports.amino = void 0;

require("./declarations");

var amino = _interopRequireWildcard(require("./amino"));

exports.amino = amino;

var _client = require("./client");

var crypto = _interopRequireWildcard(require("./crypto"));

exports.crypto = crypto;

var types = _interopRequireWildcard(require("./types"));

exports.types = types;

var utils = _interopRequireWildcard(require("./utils"));

exports.utils = utils;

var _rpc = _interopRequireDefault(require("./rpc"));

var _tx = _interopRequireDefault(require("./tx"));
